### PR TITLE
Document new deployment features

### PR DESCRIPTION
**Pull request by 0xmichalis**
Contains:
- paused deployment configs (config.spec.paused)

When a deployment config is paused, it stops being reconciled by the controller loops. This is useful for pausing a config (oc rollout pause), making multiple changes to the config and resuming (oc rollout resume) without running any new deployments in the meantime.
- cleanup policy (config.spec.revisionHistoryLimit)

If set, the controller loop will retain that number of older deployments and delete all the rest. For example, I have a dc/foo with dc.status.latestVersion=5 (meaning there are 5 deployments - 1 new and 4 old). If I set dc.spec.revisionHistoryLimit to 1 this means that only one old deployment will be kept around and the rest will be deleted. Note that only old controllers with **zero** replicas will be deleted.

``` shell
$ oc get dc
NAME       REVISION   DESIRED   CURRENT   TRIGGERED BY
database   5          1         0         config
$ oc get rc
NAME         DESIRED   CURRENT   AGE
database-1   0         0         1h
database-2   0         0         1h
database-3   0         0         1h
database-4   0         0         40m
database-5   1         1         34m
$ oc edit dc/database
# Sets revisionHistoryLimit to 1
$ oc get rc
NAME         DESIRED   CURRENT   AGE
database-4   0         0         40m
database-5   1         1         34m
```
- minReadySeconds (config.spec.minReadySeconds)

So far our deployments completed as soon as all the new pods became ready. If minReadySeconds is set, the deployer pod will not mark the deployment as Complete before this number of seconds passes for all ready pods.
- `oc rollout`

`oc rollout pause` for pausing a deployment config

`oc rollout resume` for resuming a deployment config

`oc rollout undo` for rolling back a deployment config (almost identical to `oc rollback`, lacks some of `rollback`'s flags)

`oc rollout history` for viewing any historical data available for a deployment config

Targetted for 1.3
Trello card: https://trello.com/c/Mljldox7/643-8-deployments-downstream-support-for-upstream-features

@mfojtik fyi. Still WIP.

### CODE REVIEW

File: dev_guide/deployments.adoc [hunk 1]
```diff
@@ -76,14 +74,21 @@ spec:
...
+  minReadySeconds: 0 <8>
+
 ----
 
-<1> The replication controller template named `frontend` describes a simple Ruby application.
-<2> There will be 5 replicas of `frontend` by default.
-<3> A link:#config-change-trigger[configuration change trigger] causes a new deployment to be created any time the replication controller template changes.
+<1> The pod template of the `frontend` deployment configuration describes a simple Ruby application.
+<2> There will be 5 replicas of `frontend`.
+<3> A link:#config-change-trigger[configuration change trigger] causes a new deployment to be created any time the pod template changes.
```
Comment (ahardin-rh) [inline dev_guide/deployments.adoc]: We just converted to xrefs, so we will now only be using link format for external links.
s/link:#config-change-trigger[configuration change trigger]/xref:config-change-trigger[configuration change trigger]

File: dev_guide/deployments.adoc [hunk 2]
```diff
@@ -163,61 +172,66 @@ performed using the REST API, the CLI, or the web console.
...
-This command returns the logs from the first deployment of the provided
-deployment configuration, if and only if that deployment exists (i.e., it has
-failed and has not been manually deleted or pruned).
 
 [[triggers]]
 == Triggers
 
 A deployment configuration can contain triggers, which drive the creation of
-new deployments in response to events, both inside and outside OpenShift.
+new deployments in response to events, only inside OpenShift at the moment.
```
Comment (ahardin-rh) [inline dev_guide/deployments.adoc]: s/OpenShift/{product-title}

**Comment by ahardin-rh** [comment]
@kargakis Looking good so far. Just a few minor comments from me. Also, will need a rebase. Thanks!

**Comment by kargakis** [comment]
@ahardin-rh thanks for the review. We talked with @tnguyen-rh and I will hand this pull to him for moving it forward.

@tnguyen-rh I have already made some changes, please fill free to fetch my branch and extend it. I wrote an outline of the new features in the first post and if you need any more information, you can ask @mfojtik as I will be OOO for the next three weeks and my availability will be spotty.

**Comment by ahardin-rh** [comment]
@kargakis Okay, sounds like a plan. Thanks!

**Comment by tnguyen-rh** [comment]
@kargakis Looks like this is [P2 for docs team](https://trello.com/c/ojVYSraC/49-support-upstream-deployment-features-in-deploymentconfigs-platform), which basically means we won't get to it immediately.  If you want to see it in origin soon, you can re-open this, get it to "good enough" state (w/ the help of @ahardin-rh as before), and we'll add the `to_followup` label for further downstream work.  OTOH, if you can wait, @vikram-redhat will create a BZ and assign it (maybe to me) the normal way.

**Comment by kargakis** [comment]
I would love to get these new changes in so I am back at working at this.

File: dev_guide/deployments.adoc (hunk 2 above)
Comment (kargakis) [inline dev_guide/deployments.adoc]: Changed all instances of OpenShift to {product-title} in this doc

File: dev_guide/deployments.adoc (hunk 1 above)
Comment (kargakis) [inline dev_guide/deployments.adoc]: Changed all links to use xrefs

**Comment by kargakis** [comment]
@mfojtik @smarterclayton I have started transitioning "deployment" to "replication controller" or "deployment process" where it made sense.

**Comment by ahardin-rh** [comment]
Closing this and moving over to https://github.com/openshift/openshift-docs/pull/2694 for a fresh restart

**Pull request closed without merging**